redirect/src/main/java/com/adyen/checkout/redirect/RedirectUtil.kt

```diff
@@ -107,28 +83,79 @@ object RedirectUtil {
      * @return And intent that targets either another app or a Web page.
      */
     @JvmStatic
-    fun createRedirectIntent(context: Context, uri: Uri): Intent {
-        return if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
-            createCustomTabIntent(context, uri)
-        } else {
-            if (determineResolveResult(context, uri).type == ResolveResult.Type.APPLICATION) {
-                Intent(Intent.ACTION_VIEW, uri)
-            } else {
-                createCustomTabIntent(context, uri)
+    fun launchUri(context: Context, uri: Uri) {
+        val launched = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R)
+            launchNativeApi30(context, uri)
+        else launchNativeBeforeApi30(context, uri)
+
+        if (!launched) {
+            //open in custom tabs if there's no native app for the target uri
+            val defaultColors = CustomTabColorSchemeParams.Builder()
+                .setToolbarColor(ThemeUtil.getPrimaryThemeColor(context))
+                .build()
+            try {
+                CustomTabsIntent.Builder()
+                    .setShowTitle(true)
+                    .setDefaultColorSchemeParams(defaultColors)
+                    .build().launchUrl(context, uri)
+            } catch (e: ActivityNotFoundException) {
+                // in case the device doesn't support custom tabs or doesn't support google services.
+                val browserActivityIntent = Intent()
```

Comment:
Same comment here, can we extract this to a separate method with its own try catch, etc. ?

Reply:
This 'catch' block is just for handling if there's no custom tabs support or google play services or Huawei devices. 
The 'try' and 'catch' block for it would be just to throw an exception if there's something wrong, okay?

Reply:
it will be something like this

```
private fun launchWithCustomTabs(context: Context, uri: Uri): Boolean {
        //open in custom tabs if there's no native app for the target uri
        val defaultColors = CustomTabColorSchemeParams.Builder()
            .setToolbarColor(ThemeUtil.getPrimaryThemeColor(context))
            .build()
        return try {
            CustomTabsIntent.Builder()
                .setShowTitle(true)
                .setDefaultColorSchemeParams(defaultColors)
                .build()
                .launchUrl(context, uri)
            true
        } catch (e: ActivityNotFoundException) {
            false
        }
    }

    /**
     * in case the device doesn't support custom tabs or doesn't support google services (Huawei device).
     */
    private fun launchBrowser(context: Context, uri: Uri) {
        try {
            val browserActivityIntent = Intent()
                .setAction(Intent.ACTION_VIEW)
                .addCategory(Intent.CATEGORY_BROWSABLE)
                .addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
                .setData(uri)
            context.startActivity(browserActivityIntent)
        } catch (e: ActivityNotFoundException) {
            throw ComponentException("Redirect to app failed.", e)
        }
    }
```

```
   /**
     * Creates the intent that will start the redirect.
     * @param context Any context.
     * @param uri The Uri to redirect to.
     * @return And intent that targets either another app or a Web page.
     */
    @JvmStatic
    fun launchUri(context: Context, uri: Uri) {
        val launched = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R)
            launchNativeApi30(context, uri)
        else launchNativeBeforeApi30(context, uri)

        if (!launched) {
            val customTabsLaunch = launchWithCustomTabs(context, uri)
            if (!customTabsLaunch) {
                launchBrowser(context, uri)
            }
        }
    }
```

Reply:
👌